server/publications/collections/product.js

```diff
@@ -53,42 +54,56 @@ export function findProductMedia(publicationInstance, productIds) {
  */
 Meteor.publish("Product", function (productId) {
   check(productId, Match.OptionalOrNull(String));
+
   if (!productId) {
     Logger.debug("ignoring null request on Product subscription");
     return this.ready();
   }
+
+  // verify that parent shop is ready
+  const shop = Reaction.getCurrentShop();
+  if (typeof shop !== "object") {
+    return this.ready();
+  }
+
+
   let _id;
+  let productShopId;
+
+  // Default selector. Any changes come via specific roles
   const selector = {
     isDeleted: { $in: [null, false] },
     isVisible: true
   };
-  let productShopId;
-  const shop = Reaction.getCurrentShop();
-  // verify that parent shop is ready
-  if (typeof shop !== "object") {
-    return this.ready();
+
+  // Permits admins to view both visible and invisible products.
+  if (Roles.userIsInRole(this.userId, ["owner", "admin", "createProduct"], shop._id)) {
+    selector.isVisible = {
+      $in: [true, false]
+    };
   }
 
-  // Take productShopId in order to check if user can edit this product or view its revisions
   // TODO review for REGEX / DOS vulnerabilities.
   if (productId.match(/^[23456789ABCDEFGHJKLMNPQRSTWXYZabcdefghijkmnopqrstuvwxyz]{17}$/)) {
     // selector._id = productId;
     // TODO try/catch here because we can have product handle passed by such regex
     _id = productId;
 
     const product = Products.findOne(_id);
+
     if (product) {
+      // Take productShopId in order to check if user can edit this product or view its revisions
       productShopId = product.shopId;
     } else {
       return this.ready();
     }
   } else {
-    const newSelector = {
+    const newSelector = Object.assign({}, selector, {
```

Comment:
This is the other change that fixes a broken test. Without this section, the test ensuring that invisible products are not available to non-admins is broken.

I can see how this seems unnecessary with the section above that permits admins to view invisible products, so I'm curious what the bigger picture is here if this was removed on purpose.